### ISSUE TITLE
make stable against rendering errors by using multiple tools

### ISSUE THREAD

**Issue by drtrigon** [AbdealiJK/file-metadata#37]
Mediawiki (e.g. commons) use imagemagick `convert` in order to process image media. In order to be prone against rendering errors of such tools use at least 2 (complementary) tools to render images and compare against each other.

**Comment by AbdealiLoKo** [comment]
Related pages on commons:
- https://commons.wikimedia.org/wiki/Category:PDF_files_affected_by_MediaWiki_restrictions
- https://commons.wikimedia.org/wiki/Category:Images_without_thumbnails
- https://commons.wikimedia.org/wiki/Category:Images_with_render_problem

**Comment by AbdealiLoKo** [comment]
Was thinking about this a bit more, and it seems like this is probably out of scope for file-metadata.
file-metadata aims to provide meta information about the file by using exif data or computer vision, and rendering faults (although a useful feature) would not be a part of meta information.

It would be better to focus on a single core competency, rather than delve into multiple things.
Possibly, we could close this and open a issue on phabricator "pywikibot script idea: Check render faults in mediawiki" ?

**Comment by jayvdb** [comment]
Perhaps the in-scope part is that file-metadata should distinguish between failures
- specific to one of its dependencies
- caused by the file bits being incorrect

Rather than wrapping existing exceptions from client libraries, create two tuples:

```
FileEncodingError = (dlib.xxxx.FooError, opencv.xxxx.Error, etc)
ConverterFailure = (dlib.xxxx.RefuseProcessing, opencv.xxxx.Blah, etc)
```

Then calling code can do

```
if isinstance(e, file_metadata.exceptions.FileEncodingError):
    print('file ... is broken: %s' % e)
elif isinstance(e, file_metadata.exceptions.ConverterFailure)::
    print('file ... can not be processed by file_metadata: %s' % e)
else:
    print('file failed processing by file_metadata for unknown reason: %s' %e)
```

**Comment by drtrigon** [comment]
May be the scope shifted - catimages should use any mean to get info about a file for categorization. This is one of it. So as I see it:
1. this rendering check is part of catimages
2. may be not of file-metadata but then it needs to be put into another package
3. (I don't think it is very high priority right now)

So it belongs to catimages todo or fudo (future todo) but IHMO it's not a separate bot - e.g. you don't want another bot script to download all files again just for render check.

Am 13. Juni 2016 06:15:59 MESZ, schrieb AbdealiJK notifications@github.com:

> Was thinking about this a bit more, and it seems like this is probably
> out of scope for file-metadata.
> file-metadata aims to provide meta information about the file by using
> exif data or computer vision, and rendering faults (although a useful
> feature) would not be a part of meta information.
> 
> It would be better to focus on a single core competency, rather than
> delve into multiple things.
> Possibly, we could close this and open a issue on phabricator
> "pywikibot script idea: Check render faults in mediawiki" ?
> 
> ---
> 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/AbdealiJK/file-metadata/issues/37#issuecomment-225486169

Dr. Trigon

**Comment by AbdealiLoKo** [comment]
Yep, that's what I meant - it may be potentially in catimages, but not in file-metadata.

**Comment by drtrigon** [comment]
Am 13. Juni 2016 06:27:47 MESZ, schrieb John Vandenberg notifications@github.com:

> Perhaps the in-scope part is that file-metadata should distinguish
> between failures
> - specific to one of its dependencies (1)
> - caused by the file bits being incorrect (2)

Totally agree. We need to catch errors that give use info about the file (1) and raise only the ones we cannot interprete (2 - e.g. cause by true issue).
Although (1) could get involved and might need to involve machine learning (categorization per se) and training in order to separate (1) from (2).

> Rather than wrapping existing exceptions from client libraries, create
> two tuples:
> 
> ```
> FileEncodingError = (dlib.xxxx.FooError, opencv.xxxx.Error, etc)
> ConverterFailure = (dlib.xxxx.RefuseProcessing, opencv.xxxx.Blah, etc)
> ```
> 
> Then calling code can do
> 
> ```
> if isinstance(e, file_metadata.exceptions.FileEncodingError):
>    print('file ... is broken: %s' % e)
> elif isinstance(e, file_metadata.exceptions.ConverterFailure)::
>    print('file ... can not be processed by file_metadata: %s' % e)
> else:
> print('file failed processing by file_metadata for unknown reason: %s'
> %e)
> ```

Did you mean:

isinst = lambda inst: isinstance(e, inst)

> if True not in map(isinst(file_metadata.exceptions.FileEncodingError)):
>    print('file ... is broken: %s' % e)
> elif ...

May be you can avoid the lambda.

> ---
> 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/AbdealiJK/file-metadata/issues/37#issuecomment-225487054

Dr. Trigon

**Comment by jayvdb** [comment]
No. isinstance can be given a tuple.
